odoo/fields.py

```diff
@@ -3070,6 +3075,676 @@ def _record_ids_per_res_model(self, records):
+
+        - many2one: (35, 'Bob') -> 35
+        - many2many: [(35, 'Bob'), (36, 'Alice')] -> [35, 36]
+
+        :param values_list: List of properties definition with properties value
+        :param value_key: In which dict key we need to remove the display name
+        """
+        for property_definition in values_list:
+            if not isinstance(property_definition, dict) or not property_definition.get('name'):
+                continue
+
+            property_value = property_definition.get(value_key)
+            if not property_value:
+                continue
+
+            property_type = property_definition.get('type')
+
+            if property_type == 'many2one' and has_list_types(property_value, [int, (str, NoneType)]):
+                property_definition[value_key] = property_value[0]
+
+            elif property_type == 'many2many':
+                if is_list_of(property_value, (list, tuple)):
+                    # [(35, 'Admin'), (36, 'Demo')] -> [35, 36]
+                    property_definition[value_key] = [
+                        many2many_value[0]
+                        for many2many_value in property_value
+                    ]
+
+    @classmethod
+    def _parse_json_types(cls, values_list, env, check_existence=True):
+        """Parse the value stored in the JSON.
+
+        Check for records existence, if we removed a selection option, ...
+        Modify in place "values_list".
+
+        :param values_list: List of properties definition and values
+        :param env: environment
+        """
+        for property_definition in values_list:
+            property_value = property_definition.get('value')
+            property_type = property_definition.get('type')
+            res_model = property_definition.get('comodel')
+
+            if property_type not in cls.ALLOWED_TYPES:
+                raise ValueError(f'Wrong property type {property_type!r}')
+
+            if property_type == 'boolean':
+                # E.G. convert zero to False
+                property_value = bool(property_value)
+
+            elif property_type == 'char' and not isinstance(property_value, str):
+                property_value = False
+
+            elif property_value and property_type == 'selection':
+                # check if the selection option still exists
+                options = property_definition.get('selection') or []
+                options = [option[0] for option in options if option and len(option) == 2]
+                if property_value not in options:
+                    # maybe the option has been removed on the parent
+                    property_value = False
+
+            elif property_value and property_type == 'tags':
+                # remove all tags that are not defined on the parent
+                all_tags = [tag[0] for tag in property_definition.get('tags') or []]
+                property_value = [tag for tag in property_value if tag in all_tags]
```

Comment:
avoid O(n2) here too, someone is bound to come up with a huge list tags at some point ^^

```suggestion
                all_tags = set(tag[0] for tag in property_definition.get('tags') or [])
```

Reply:
You can even use a set comprehension for that :wink: (I think you save the generator's overhead.)
```suggestion
                all_tags = {tag[0] for tag in property_definition.get('tags') or ()}
                property_value = [tag for tag in property_value if tag in all_tags]
```